src/ledger/LedgerStateSnapshot.h

```diff
@@ -158,12 +165,43 @@ class LedgerSnapshot : public NonMovableOrCopyable
         return mGetter->getAccount(header, tx, AccountID);
     }
     LedgerEntryWrapper load(LedgerKey const& key) const;
+};
+
+// A subclass of LedgerSnapshot that contains additional snapshots necessary for
+// transaction validation.
+class ExtendedLedgerSnapshot : public LedgerSnapshot
+{
+  public:
+    // Generate a snapshot from the most recently closed ledger. The resulting
+    // `ExtendedLedgerSnapshot` cannot be used for transaction application.
+    explicit ExtendedLedgerSnapshot(Application const& app);
+
+    // Generate a snapshot from a specific AbstractLedgerTxn. Set `forApply` to
+    // `true` to use this snapshot in transaction application.
+    ExtendedLedgerSnapshot(AbstractLedgerTxn& ltx, AppConnector const& app,
+                           bool forApply);
+
+    // Generate a snapshot from a specific `AbstractLedgerTxn`, copying
+    // additional snapshot data from `outer`. This is intended only for use with
+    // `executeWithMaybeInnerSnapshot`.
+    ExtendedLedgerSnapshot(AbstractLedgerTxn& ltx,
+                           ExtendedLedgerSnapshot const& outer);
+    virtual ~ExtendedLedgerSnapshot() = default;
 
     // Execute a function with a nested snapshot, if supported. This is needed
     // to support the replay of old buggy protocols (<8), see
     // `TransactionFrame::loadSourceAccount`
     void executeWithMaybeInnerSnapshot(
-        std::function<void(LedgerSnapshot const&)> f) const;
-};
+        std::function<void(ExtendedLedgerSnapshot const&)> f) const;
 
+    // Getters for the additional snapshots
+    Config const& getConfig() const;
+    SorobanNetworkConfig const& getSorobanNetworkConfig() const;
+    uint32_t getCurrentProtocolVersion() const;
+
+  private:
+    std::shared_ptr<const Config> const mConfig;
```

Comment:
mConfig doesn't belong to a snapshot, since it's `const` right? Typically, we should be able to access config from either mApp or mAppConnector.

Reply:
Part of the design for this change was to make it impossible for `checkValid` to interact with other subsystems directly (outside of snapshots). I removed `AppConnector` from the validation flow entirely to ensure this was the case. As a result, I folded the config into `ExtendedLedgerSnapshot`, but if we don't like it there than I can manually pass it through through validation flow.